{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n",
      "2024-02-28 19:26:39,927\tINFO util.py:159 -- Missing packages: ['ipywidgets']. Run `pip install -U ipywidgets`, then restart the notebook server for rich notebook output.\n",
      "2024-02-28 19:26:40,047\tINFO util.py:159 -- Missing packages: ['ipywidgets']. Run `pip install -U ipywidgets`, then restart the notebook server for rich notebook output.\n"
     ]
    }
   ],
   "source": [
    "# Imports\n",
    "try:\n",
    "    from tqdm import tqdm\n",
    "except ImportError:\n",
    "    def tqdm(iterable):\n",
    "        return iterable\n",
    "\n",
    "import torch\n",
    "import pathlib\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import utils\n",
    "import config\n",
    "# sys.path.append('./baseline_src')\n",
    "\n",
    "import ray \n",
    "from ray.tune.search import ConcurrencyLimiter\n",
    "from ray.tune.search.bayesopt import BayesOptSearch\n",
    "from ray import tune\n",
    "import json\n",
    "from statistics import mean\n",
    "# from config import DynamicGraphTemporalSignal_custom\n",
    "from ray.tune.search import ConcurrencyLimiter\n",
    "from ray.tune.search.bayesopt import BayesOptSearch\n",
    "import joblib\n",
    "import random\n",
    "from torch.utils.data import DataLoader, TensorDataset\n",
    "import joblib\n",
    "import random\n",
    "\n",
    "torch.manual_seed(0)\n",
    "random.seed(0)\n",
    "np.random.seed(0)\n",
    "\n",
    "device = 'cpu'\n",
    "\n",
    "num_columns = ['precipitation_intensity','temperature','humidity','average_speed','extreme_congestion','past_week_load_21', # NODE\n",
    "               'distance_traveled',\n",
    "            #    'total_traveled','median_income_last12months','white_pct','black_pct', # EDGE\n",
    "            #    'hispanic_pct','public_transit_pct','pct_public_transit_for_work' # EDGE\n",
    "               ]\n",
    "ohe_columns = ['route_id','is_holiday','is_weekend','route_direction_name']\n",
    "cat_columns = ['time_window_cat','stop_sequence_cat','year_cat','month_cat']\n",
    "\n",
    "TARGET = 'trip_load_4_bin_transit'\n",
    "\n",
    "REQ_COLS = num_columns + ohe_columns + cat_columns + [TARGET]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def train_one_epoch(model,train_loader,focal_loss,optimizer):\n",
    "    model.train()\n",
    "\n",
    "    time = 0\n",
    "    loss = 0\n",
    "    for inputs, targets in train_loader:\n",
    "        inputs.to(device)\n",
    "        targets.to(device)\n",
    "        outputs = model(inputs)\n",
    "        # targets = np.argmax(targets,axis=-1)\n",
    "        loss = loss + focal_loss(outputs, targets)\n",
    "        time += 1\n",
    "    loss = loss / (time+1) #Getting the mean loss\n",
    "    loss.backward()\n",
    "    optimizer.step()\n",
    "    optimizer.zero_grad()\n",
    "\n",
    "    return loss\n",
    "\n",
    "def validate_one_epoch(model,validation_loader,focal_loss,optimizer):\n",
    "    model.eval()\n",
    "    time = 0\n",
    "    loss = 0\n",
    "    for inputs, targets in validation_loader:\n",
    "        inputs.to(device)\n",
    "        targets.to(device)\n",
    "        outputs = model(inputs)\n",
    "        loss = loss + focal_loss(outputs, targets)\n",
    "        time += 1\n",
    "    loss = loss / (time+1) #Getting the mean loss\n",
    "\n",
    "    return loss"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def train_and_evaluate_model(model, epochs, patience, train_loader, validation_loader, focal_loss, optimizer):\n",
    "\n",
    "    early_stopper = utils.EarlyStopper(patience=patience, min_delta=0)\n",
    "\n",
    "    history = {\n",
    "        'epoch':[],\n",
    "        'train_loss':[],\n",
    "        'validation_loss':[],\n",
    "    }    # Training loop\n",
    "\n",
    "    for epoch in tqdm(range(epochs)):\n",
    "        train_loss = train_one_epoch(model,train_loader,focal_loss,optimizer)\n",
    "        validation_loss = validate_one_epoch(model,validation_loader,focal_loss,optimizer)\n",
    "\n",
    "        val_loss_to_monitor = validation_loss\n",
    "\n",
    "        history['epoch'].append(epoch+1)\n",
    "        history['train_loss'].append(float(train_loss.data))\n",
    "        history['validation_loss'].append(float(validation_loss.data))\n",
    "        \n",
    "        if early_stopper.early_stop(val_loss_to_monitor):     \n",
    "            # print(f\"Stopping early! {validation_loss}, {epoch}\")      \n",
    "            print(f\"Stopping early! {train_loss}, {epoch}\")      \n",
    "            break\n",
    "\n",
    "    return model,history\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# %%\n",
    "def train_model(hyperparams):\n",
    "\n",
    "    #Hyperparams\n",
    "    epochs = int(hyperparams['epochs'])\n",
    "    patience = int(hyperparams['patience'])\n",
    "    learning_rate = hyperparams['learning_rate']\n",
    "\n",
    "    batch_size = int(hyperparams['batch_size'])\n",
    "    num_dense_layers = int(hyperparams['num_dense_layers'])\n",
    "    hidden_neurons = int(hyperparams['hidden_neurons'])\n",
    "    dropout_rate = hyperparams['dropout_rate']\n",
    "    random_seed = int(hyperparams['random_seed'])\n",
    "    num_classes = hyperparams['route_info']['num_classes']\n",
    "\n",
    "    route_id = hyperparams['route_info']['route_id']\n",
    "\n",
    "    if route_id == 999:\n",
    "        route_id = ''\n",
    "\n",
    "    cv_train_loss = []\n",
    "    cv_val_loss = []\n",
    "\n",
    "    fold_no = 0\n",
    "    for fold_no in range(config.CROSS_FOLD_COUNT):\n",
    "        # LOADING THE DATA \n",
    "        data_path = f'{config.path_to_main_folder}/data/graph_data/{config.time_window_ver}_time_window_static_graph/experiment_{config.exp_type}/'\n",
    "        train_fp = f\"{data_path}/{route_id}/train_data_fold_{fold_no}/train_data_fold_{fold_no}_processed_data.parquet\"\n",
    "        val_fp = f\"{data_path}/{route_id}/val_data_fold_{fold_no}/val_data_fold_{fold_no}_processed_data.parquet\"\n",
    "\n",
    "        train_data = pd.read_parquet(train_fp)\n",
    "        val_data = pd.read_parquet(val_fp)\n",
    "        ohe = joblib.load(f'{data_path}/{route_id}/train_data_fold_{fold_no}/ohe.joblib')\n",
    "\n",
    "        train_data = train_data[:1000]\n",
    "        val_data = train_data\n",
    "\n",
    "        train_data = train_data[REQ_COLS]\n",
    "        val_data = val_data[REQ_COLS]\n",
    "                \n",
    "        data_to_scale = train_data[ohe_columns]\n",
    "        train_data[ohe.get_feature_names_out()] = ohe.transform(data_to_scale).toarray()\n",
    "\n",
    "        data_to_scale = val_data[ohe_columns]\n",
    "        val_data[ohe.get_feature_names_out()] = ohe.transform(data_to_scale).toarray()\n",
    "\n",
    "        train_data.drop(ohe_columns,axis=1,inplace=True)\n",
    "        val_data.drop(ohe_columns,axis=1,inplace=True)\n",
    "\n",
    "        # GETTING THE the class weights used for focal loss\n",
    "        y_train = np.array(train_data[TARGET])\n",
    "        class_counts = np.bincount(y_train)\n",
    "        total_samples = len(y_train)\n",
    "        class_weights = []\n",
    "        for count in class_counts:\n",
    "            weight = 1 / (count / total_samples)\n",
    "            class_weights.append(weight)\n",
    "\n",
    "        X_train = train_data[num_columns+cat_columns+list(ohe.get_feature_names_out())] \n",
    "        y_train = train_data[TARGET]\n",
    "\n",
    "        X_val = val_data[num_columns+cat_columns+list(ohe.get_feature_names_out())]\n",
    "        y_val = val_data[TARGET]\n",
    "\n",
    "        input_size = X_train.shape[-1]\n",
    "\n",
    "        train_dataset = TensorDataset(torch.tensor(np.array(X_train)).float() , torch.tensor(np.array(y_train))) # Flatten y for compatibility with DataLoader\n",
    "        train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)\n",
    "\n",
    "        validation_dataset = TensorDataset(torch.tensor(np.array(X_val)).float() , torch.tensor(np.array(y_val))) # Flatten y for compatibility with DataLoader\n",
    "        validation_loader = DataLoader(validation_dataset, batch_size=batch_size, shuffle=True)\n",
    "\n",
    "        torch.manual_seed(random_seed)\n",
    "        random.seed(random_seed)\n",
    "        np.random.seed(random_seed)\n",
    "        # CREATE THE MODEL\n",
    "        model = utils.MLPModel(input_size, num_classes, hyperparams)\n",
    "        model = model.to(device)\n",
    "        # Setting up the optimizer\n",
    "        optimizer = torch.optim.Adam(model.parameters(), lr=learning_rate)\n",
    "        # Setting the class weights to a tensor\n",
    "        class_weights = torch.FloatTensor(class_weights).to(device)\n",
    "        # Setting the focal loss according to the class weights \n",
    "        focal_loss = torch.hub.load(\n",
    "            'adeelh/pytorch-multi-class-focal-loss',\n",
    "            model='FocalLoss',\n",
    "            alpha=class_weights,\n",
    "            gamma=2,\n",
    "            reduction='mean',\n",
    "            force_reload=False\n",
    "        ).to(device)\n",
    "\n",
    "        # TRAIN THE MODEL \n",
    "        model, history = train_and_evaluate_model(model, epochs, patience, train_dataset, validation_dataset, focal_loss, optimizer)\n",
    "        index_val = history['validation_loss'].index(min(history['validation_loss']))\n",
    "        \n",
    "        #Storing all the data\n",
    "        cv_train_loss.append(history['train_loss'][index_val])\n",
    "        cv_val_loss.append(history['validation_loss'][index_val])\n",
    "\n",
    "        # Call back to save the best model till now\n",
    "        # route_id = hyperparams['route_info']['route_id']\n",
    "\n",
    "        print(f'Training for fold {fold_no} ...')\n",
    "        fold_no = fold_no+1\n",
    "\n",
    "    cv_val_loss_mean = mean(cv_val_loss)\n",
    "\n",
    "    tune.report(\n",
    "        cv_val_loss_mean=cv_val_loss_mean,\n",
    "        cv_train_loss=cv_train_loss,\n",
    "        cv_val_loss=cv_val_loss,\n",
    "    )\n",
    "\n",
    "    return model,history\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# %%\n",
    "def train_tune(hyperparams):\n",
    "    # Start Ray experiment and pass the `hyperparams` to the `train` function\n",
    "    model,history = train_model(hyperparams)\n",
    "    print(\"Saving the model and history\")\n",
    "\n",
    "    def save_model_and_history_callback(model, history, hyperparams):\n",
    "            epochs = int(hyperparams['epochs'])\n",
    "            patience = int(hyperparams['patience'])\n",
    "            learning_rate = hyperparams['learning_rate']\n",
    "            batch_size = int(hyperparams['batch_size'])\n",
    "            num_dense_layers = int(hyperparams['num_dense_layers'])\n",
    "            hidden_neurons = int(hyperparams['hidden_neurons'])\n",
    "            dropout_rate = hyperparams['dropout_rate']\n",
    "            random_seed = int(hyperparams['random_seed'])\n",
    "            num_classes = hyperparams['route_info']['num_classes']\n",
    "            route_id = hyperparams['route_info']['route_id']    \n",
    "\n",
    "            hp_filename = f'{epochs}_{patience}_lr_{learning_rate}_bs_{batch_size}_ndl_{num_dense_layers}_hn_{hidden_neurons}_rs_{random_seed}_do_{dropout_rate}'\n",
    "\n",
    "            folder = f'../models/{hp_filename}'\n",
    "            pathlib.Path(folder).mkdir(parents=True, exist_ok=True) \n",
    "\n",
    "            # Generate a unique file name for each trial\n",
    "            file_name = f\"{folder}/model_{hp_filename}.pt\"    \n",
    "            # Save the model to disk\n",
    "            torch.save({\n",
    "                'epoch': epochs,\n",
    "                'model_state_dict': model.state_dict(),\n",
    "            }, file_name)\n",
    "            \n",
    "            # Save the history to disk\n",
    "            history_file_name = f\"{folder}/history_{hp_filename}.json\"\n",
    "            with open(history_file_name, 'w') as f:\n",
    "                json.dump(history, f)\n",
    "                \n",
    "    save_model_and_history_callback(model,history,hyperparams)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# %%\n",
    "# Function for ray hyperparameter tuning \n",
    "def ray_hyper_parameter_tune(route_info):\n",
    "\n",
    "    # TODO: FIX BELOW \n",
    "    hyperparams = {\n",
    "        'epochs':   tune.quniform(100, 600, 1),\n",
    "        'patience': tune.quniform(50,150,1),\n",
    "        'learning_rate' : tune.uniform(0.0001, 0.1),\n",
    "        'batch_size' : tune.quniform(64,128,1),\n",
    "        'num_dense_layers': tune.quniform(1, 5, 1),\n",
    "        'hidden_neurons': tune.quniform(1,256,1),\n",
    "        'dropout_rate' : tune.uniform(0,0.5),\n",
    "        'random_seed': tune.quniform(0, 300, 1),\n",
    "        'route_info': route_info\n",
    "    }\n",
    "\n",
    "    dir_fp = f\"ray_tune/hyperparams_output/STL_{config.version}/\"\n",
    "    local_dir = f'../{dir_fp}'\n",
    "    pathlib.Path(local_dir).mkdir(parents=True, exist_ok=True) \n",
    "\n",
    "    ray.init()\n",
    "\n",
    "    path_to_local_dir = f'{config.path_to_main_folder}/{dir_fp}'\n",
    "\n",
    "    algo = BayesOptSearch(utility_kwargs={\"kind\": \"ucb\", \"kappa\": 2.5, \"xi\": 0.0})\n",
    "    algo = ConcurrencyLimiter(algo, max_concurrent=3)\n",
    "   \n",
    "    analysis = tune.run(\n",
    "            #  \n",
    "            # If you want to use other parameters, you can add them below like tune.with_parameters(train_tune,train_dataset=train_dataset,validation_dataset=validation_dataset) . Function: def train_tune(hyperparams,train_dataset,validation_dataset)\n",
    "            tune.with_parameters(train_tune),\n",
    "            config=hyperparams,\n",
    "            resources_per_trial={\"cpu\": 4},\n",
    "            metric=\"cv_val_loss_mean\",\n",
    "            mode=\"min\",\n",
    "            search_alg=algo,\n",
    "            num_samples=config.RAY_NUM_SAMPLES,\n",
    "            # max_concurrent_trials=5,\n",
    "            local_dir= f'{path_to_local_dir}',\n",
    "            name=f\"experiment_STL_{config.version}_{route_info['route_id']}\",\n",
    "            max_failures=1,\n",
    "            raise_on_failed_trial=False\n",
    "        )\n",
    "\n",
    "    # Stop Ray\n",
    "    ray.shutdown()\n",
    "\n",
    "    temp = analysis.dataframe()   \n",
    "    hyperparams_path = f'{local_dir}'\n",
    "    pathlib.Path(hyperparams_path).mkdir(parents=True, exist_ok=True) \n",
    "    temp.to_csv(f\"{hyperparams_path}/RAY_RESULTS_STL_{config.version}_ROUTE_{route_info['route_id']}.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# route_info = {} \n",
    "\n",
    "# route_info['route_id'] = 999\n",
    "# route_info['num_classes'] = 5\n",
    "\n",
    "# hyperparams = {\n",
    "#         'epochs':   10,\n",
    "#         'patience': 3,\n",
    "#         'learning_rate' : 0.01,\n",
    "#         'GNN_hidden_channels': 36,\n",
    "#         'gru_layers': 2,\n",
    "#         'chebyshev_filter': 3,\n",
    "#         'linear_layer_count': 3,\n",
    "#         'linear_hidden_channels': 20,\n",
    "#         'dropout_rate' : 0.1,\n",
    "#         'route_info': route_info\n",
    "#     }\n",
    "# model,history = train_model(hyperparams)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-02-28 19:26:41,981\tINFO worker.py:1612 -- Started a local Ray instance. View the dashboard at \u001b[1m\u001b[32m127.0.0.1:8265 \u001b[39m\u001b[22m\n",
      "2024-02-28 19:26:42,694\tINFO tune.py:657 -- [output] This uses the legacy output and progress reporter, as Jupyter notebooks are not supported by the new engine, yet. For more information, please see https://github.com/ray-project/ray/issues/36949\n",
      "/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/tune/tune.py:258: UserWarning: Passing a `local_dir` is deprecated and will be removed in the future. Pass `storage_path` instead or set the `RAY_AIR_LOCAL_CACHE_DIR` environment variable instead.\n",
      "  warnings.warn(\n",
      "2024-02-28 19:26:42,704\tWARNING bayesopt_search.py:421 -- BayesOpt search does not support quantization. Dropped quantization.\n",
      "2024-02-28 19:26:42,704\tWARNING bayesopt_search.py:431 -- BayesOpt does not support specific sampling methods. The Uniform sampler will be dropped.\n",
      "2024-02-28 19:26:42,704\tWARNING bayesopt_search.py:421 -- BayesOpt search does not support quantization. Dropped quantization.\n",
      "2024-02-28 19:26:42,705\tWARNING bayesopt_search.py:431 -- BayesOpt does not support specific sampling methods. The Uniform sampler will be dropped.\n",
      "2024-02-28 19:26:42,705\tWARNING bayesopt_search.py:421 -- BayesOpt search does not support quantization. Dropped quantization.\n",
      "2024-02-28 19:26:42,705\tWARNING bayesopt_search.py:431 -- BayesOpt does not support specific sampling methods. The Uniform sampler will be dropped.\n",
      "2024-02-28 19:26:42,705\tWARNING bayesopt_search.py:421 -- BayesOpt search does not support quantization. Dropped quantization.\n",
      "2024-02-28 19:26:42,705\tWARNING bayesopt_search.py:431 -- BayesOpt does not support specific sampling methods. The Uniform sampler will be dropped.\n",
      "2024-02-28 19:26:42,705\tWARNING bayesopt_search.py:421 -- BayesOpt search does not support quantization. Dropped quantization.\n",
      "2024-02-28 19:26:42,706\tWARNING bayesopt_search.py:431 -- BayesOpt does not support specific sampling methods. The Uniform sampler will be dropped.\n",
      "2024-02-28 19:26:42,706\tWARNING bayesopt_search.py:421 -- BayesOpt search does not support quantization. Dropped quantization.\n",
      "2024-02-28 19:26:42,706\tWARNING bayesopt_search.py:431 -- BayesOpt does not support specific sampling methods. The Uniform sampler will be dropped.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div class=\"tuneStatus\">\n",
       "  <div style=\"display: flex;flex-direction: row\">\n",
       "    <div style=\"display: flex;flex-direction: column;\">\n",
       "      <h3>Tune Status</h3>\n",
       "      <table>\n",
       "<tbody>\n",
       "<tr><td>Current time:</td><td>2024-02-28 19:26:42</td></tr>\n",
       "<tr><td>Running for: </td><td>00:00:00.13        </td></tr>\n",
       "<tr><td>Memory:      </td><td>20.9/36.0 GiB      </td></tr>\n",
       "</tbody>\n",
       "</table>\n",
       "    </div>\n",
       "    <div class=\"vDivider\"></div>\n",
       "    <div class=\"systemInfo\">\n",
       "      <h3>System Info</h3>\n",
       "      Using FIFO scheduling algorithm.<br>Logical resource usage: 4.0/12 CPUs, 0/0 GPUs\n",
       "    </div>\n",
       "    \n",
       "  </div>\n",
       "  <div class=\"hDivider\"></div>\n",
       "  <div class=\"trialStatus\">\n",
       "    <h3>Trial Status</h3>\n",
       "    <table>\n",
       "<thead>\n",
       "<tr><th>Trial name         </th><th>status  </th><th>loc  </th><th style=\"text-align: right;\">  batch_size</th><th style=\"text-align: right;\">  dropout_rate</th><th style=\"text-align: right;\">  epochs</th><th style=\"text-align: right;\">  hidden_neurons</th><th style=\"text-align: right;\">  learning_rate</th><th style=\"text-align: right;\">  num_dense_layers</th><th style=\"text-align: right;\">  patience</th><th style=\"text-align: right;\">  random_seed</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr><td>train_tune_2b9e5c13</td><td>PENDING </td><td>     </td><td style=\"text-align: right;\">     87.9706</td><td style=\"text-align: right;\">      0.475357</td><td style=\"text-align: right;\"> 465.997</td><td style=\"text-align: right;\">         153.658</td><td style=\"text-align: right;\">      0.0156863</td><td style=\"text-align: right;\">           1.62398</td><td style=\"text-align: right;\">   55.8084</td><td style=\"text-align: right;\">      259.853</td></tr>\n",
       "</tbody>\n",
       "</table>\n",
       "  </div>\n",
       "</div>\n",
       "<style>\n",
       ".tuneStatus {\n",
       "  color: var(--jp-ui-font-color1);\n",
       "}\n",
       ".tuneStatus .systemInfo {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "}\n",
       ".tuneStatus td {\n",
       "  white-space: nowrap;\n",
       "}\n",
       ".tuneStatus .trialStatus {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "}\n",
       ".tuneStatus h3 {\n",
       "  font-weight: bold;\n",
       "}\n",
       ".tuneStatus .hDivider {\n",
       "  border-bottom-width: var(--jp-border-width);\n",
       "  border-bottom-color: var(--jp-border-color0);\n",
       "  border-bottom-style: solid;\n",
       "}\n",
       ".tuneStatus .vDivider {\n",
       "  border-left-width: var(--jp-border-width);\n",
       "  border-left-color: var(--jp-border-color0);\n",
       "  border-left-style: solid;\n",
       "  margin: 0.5em 1em 0.5em 1em;\n",
       "}\n",
       "</style>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-02-28 19:26:44,230\tWARNING worker.py:2006 -- Traceback (most recent call last):\n",
      "  File \"python/ray/_raylet.pyx\", line 1418, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1498, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1424, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1364, in ray._raylet.execute_task.function_executor\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 726, in actor_method_executor\n",
      "    return method(__ray_actor, *args, **kwargs)\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 638, in temporary_actor_method\n",
      "    raise RuntimeError(\n",
      "RuntimeError: The actor with name ImplicitFunc failed to import on the worker. This may be because needed library dependencies are not installed in the worker environment:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 677, in _load_actor_class_from_gcs\n",
      "    actor_class = pickle.loads(pickled_class)\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/cloudpickle/cloudpickle.py\", line 665, in subimport\n",
      "    __import__(name)\n",
      "ModuleNotFoundError: No module named 'src'\n",
      "\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"python/ray/_raylet.pyx\", line 1787, in ray._raylet.task_execution_handler\n",
      "  File \"python/ray/_raylet.pyx\", line 1684, in ray._raylet.execute_task_with_cancellation_handler\n",
      "  File \"python/ray/_raylet.pyx\", line 1366, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1367, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1583, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 824, in ray._raylet.store_task_errors\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 638, in temporary_actor_method\n",
      "    raise RuntimeError(\n",
      "RuntimeError: The actor with name ImplicitFunc failed to import on the worker. This may be because needed library dependencies are not installed in the worker environment:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 677, in _load_actor_class_from_gcs\n",
      "    actor_class = pickle.loads(pickled_class)\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/cloudpickle/cloudpickle.py\", line 665, in subimport\n",
      "    __import__(name)\n",
      "ModuleNotFoundError: No module named 'src'\n",
      "\n",
      "An unexpected internal error occurred while the worker was executing a task.\n",
      "2024-02-28 19:26:44,232\tWARNING worker.py:2006 -- A worker died or was killed while executing a task by an unexpected system error. To troubleshoot the problem, check the logs for the dead worker. RayTask ID: ffffffffffffffffadf12ac9a0077d3ee7a9aa7101000000 Worker ID: b639f9261241cef8a19a09a471053c20949d5fc6f012e3c173fef4c9 Node ID: efc15fd61ccb1fcdfd79f1e80cc4e7e60403d12d4dea718b61bee81e Worker IP address: 127.0.0.1 Worker port: 60258 Worker PID: 83511 Worker exit type: SYSTEM_ERROR Worker exit detail: Worker exits unexpectedly. Worker exits with an exit code None.\n",
      " Traceback (most recent call last):\n",
      "  File \"python/ray/_raylet.pyx\", line 1418, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1498, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1424, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1364, in ray._raylet.execute_task.function_executor\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 726, in actor_method_executor\n",
      "    return method(__ray_actor, *args, **kwargs)\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 638, in temporary_actor_method\n",
      "    raise RuntimeError(\n",
      "RuntimeError: The actor with name ImplicitFunc failed to import on the worker. This may be because needed library dependencies are not installed in the worker environment:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 677, in _load_actor_class_from_gcs\n",
      "    actor_class = pickle.loads(pickled_class)\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/cloudpickle/cloudpickle.py\", line 665, in subimport\n",
      "    __import__(name)\n",
      "ModuleNotFoundError: No module named 'src'\n",
      "\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"python/ray/_raylet.pyx\", line 1787, in ray._raylet.task_execution_handler\n",
      "  File \"python/ray/_raylet.pyx\", line 1684, in ray._raylet.execute_task_with_cancellation_handler\n",
      "  File \"python/ray/_raylet.pyx\", line 1366, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1367, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1583, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 824, in ray._raylet.store_task_errors\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 638, in temporary_actor_method\n",
      "    raise RuntimeError(\n",
      "RuntimeError: The actor with name ImplicitFunc failed to import on the worker. This may be because needed library dependencies are not installed in the worker environment:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 677, in _load_actor_class_from_gcs\n",
      "    actor_class = pickle.loads(pickled_class)\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/cloudpickle/cloudpickle.py\", line 665, in subimport\n",
      "    __import__(name)\n",
      "ModuleNotFoundError: No module named 'src'\n",
      "\n",
      "An unexpected internal error occurred while the worker was executing a task.\n",
      "2024-02-28 19:26:44,233\tERROR tune_controller.py:911 -- Trial task failed for trial train_tune_2b9e5c13\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/air/execution/_internal/event_manager.py\", line 110, in resolve_future\n",
      "    result = ray.get(future)\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/auto_init_hook.py\", line 24, in auto_init_wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/client_mode_hook.py\", line 103, in wrapper\n",
      "    return func(*args, **kwargs)\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/worker.py\", line 2495, in get\n",
      "    raise value\n",
      "ray.exceptions.RayActorError: The actor died unexpectedly before finishing this task.\n",
      "\tclass_name: ImplicitFunc\n",
      "\tactor_id: adf12ac9a0077d3ee7a9aa7101000000\n",
      "\tpid: 83511\n",
      "\tnamespace: 8e0d37a8-9651-4e34-a440-6bd4d070a181\n",
      "\tip: 127.0.0.1\n",
      "The actor is dead because its worker process has died. Worker exit type: SYSTEM_ERROR Worker exit detail: Worker exits unexpectedly. Worker exits with an exit code None.\n",
      " Traceback (most recent call last):\n",
      "  File \"python/ray/_raylet.pyx\", line 1418, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1498, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1424, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1364, in ray._raylet.execute_task.function_executor\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 726, in actor_method_executor\n",
      "    return method(__ray_actor, *args, **kwargs)\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 638, in temporary_actor_method\n",
      "    raise RuntimeError(\n",
      "RuntimeError: The actor with name ImplicitFunc failed to import on the worker. This may be because needed library dependencies are not installed in the worker environment:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 677, in _load_actor_class_from_gcs\n",
      "    actor_class = pickle.loads(pickled_class)\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/cloudpickle/cloudpickle.py\", line 665, in subimport\n",
      "    __import__(name)\n",
      "ModuleNotFoundError: No module named 'src'\n",
      "\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"python/ray/_raylet.pyx\", line 1787, in ray._raylet.task_execution_handler\n",
      "  File \"python/ray/_raylet.pyx\", line 1684, in ray._raylet.execute_task_with_cancellation_handler\n",
      "  File \"python/ray/_raylet.pyx\", line 1366, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1367, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1583, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 824, in ray._raylet.store_task_errors\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 638, in temporary_actor_method\n",
      "    raise RuntimeError(\n",
      "RuntimeError: The actor with name ImplicitFunc failed to import on the worker. This may be because needed library dependencies are not installed in the worker environment:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 677, in _load_actor_class_from_gcs\n",
      "    actor_class = pickle.loads(pickled_class)\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/cloudpickle/cloudpickle.py\", line 665, in subimport\n",
      "    __import__(name)\n",
      "ModuleNotFoundError: No module named 'src'\n",
      "\n",
      "An unexpected internal error occurred while the worker was executing a task.\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m Worker exits with an exit code None.\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m  Traceback (most recent call last):\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m   File \"python/ray/_raylet.pyx\", line 1418, in ray._raylet.execute_task\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m   File \"python/ray/_raylet.pyx\", line 1498, in ray._raylet.execute_task\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m   File \"python/ray/_raylet.pyx\", line 1424, in ray._raylet.execute_task\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m   File \"python/ray/_raylet.pyx\", line 1364, in ray._raylet.execute_task.function_executor\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m   File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 726, in actor_method_executor\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m     return method(__ray_actor, *args, **kwargs)\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m   File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 638, in temporary_actor_method\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m     raise RuntimeError(\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m RuntimeError: The actor with name ImplicitFunc failed to import on the worker. This may be because needed library dependencies are not installed in the worker environment:\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m \n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m Traceback (most recent call last):\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m   File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 677, in _load_actor_class_from_gcs\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m     actor_class = pickle.loads(pickled_class)\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m   File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/cloudpickle/cloudpickle.py\", line 665, in subimport\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m     __import__(name)\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m ModuleNotFoundError: No module named 'src'\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m \n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m \n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m During handling of the above exception, another exception occurred:\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m \n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m Traceback (most recent call last):\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m   File \"python/ray/_raylet.pyx\", line 1787, in ray._raylet.task_execution_handler\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m   File \"python/ray/_raylet.pyx\", line 1684, in ray._raylet.execute_task_with_cancellation_handler\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m   File \"python/ray/_raylet.pyx\", line 1366, in ray._raylet.execute_task\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m   File \"python/ray/_raylet.pyx\", line 1367, in ray._raylet.execute_task\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m   File \"python/ray/_raylet.pyx\", line 1583, in ray._raylet.execute_task\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m   File \"python/ray/_raylet.pyx\", line 824, in ray._raylet.store_task_errors\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m   File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 638, in temporary_actor_method\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m     raise RuntimeError(\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m RuntimeError: The actor with name ImplicitFunc failed to import on the worker. This may be because needed library dependencies are not installed in the worker environment:\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m \n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m Traceback (most recent call last):\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m   File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 677, in _load_actor_class_from_gcs\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m     actor_class = pickle.loads(pickled_class)\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m   File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/cloudpickle/cloudpickle.py\", line 665, in subimport\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m     __import__(name)\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m ModuleNotFoundError: No module named 'src'\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m \n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m An unexpected internal error occurred while the worker was executing a task.\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m Traceback (most recent call last):\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m   File \"python/ray/_raylet.pyx\", line 1418, in ray._raylet.execute_task\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m   File \"python/ray/_raylet.pyx\", line 1498, in ray._raylet.execute_task\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m   File \"python/ray/_raylet.pyx\", line 1424, in ray._raylet.execute_task\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m   File \"python/ray/_raylet.pyx\", line 1364, in ray._raylet.execute_task.function_executor\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m   File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 726, in actor_method_executor\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m     return method(__ray_actor, *args, **kwargs)\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m   File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 638, in temporary_actor_method\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m     raise RuntimeError(\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m RuntimeError: The actor with name ImplicitFunc failed to import on the worker. This may be because needed library dependencies are not installed in the worker environment:\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m \n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m Traceback (most recent call last):\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m   File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 677, in _load_actor_class_from_gcs\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m     actor_class = pickle.loads(pickled_class)\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m   File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/cloudpickle/cloudpickle.py\", line 665, in subimport\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m     __import__(name)\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m ModuleNotFoundError: No module named 'src'\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m \n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m \n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m During handling of the above exception, another exception occurred:\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m \n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m Traceback (most recent call last):\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m   File \"python/ray/_raylet.pyx\", line 1787, in ray._raylet.task_execution_handler\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m   File \"python/ray/_raylet.pyx\", line 1684, in ray._raylet.execute_task_with_cancellation_handler\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m   File \"python/ray/_raylet.pyx\", line 1366, in ray._raylet.execute_task\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m   File \"python/ray/_raylet.pyx\", line 1367, in ray._raylet.execute_task\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m   File \"python/ray/_raylet.pyx\", line 1583, in ray._raylet.execute_task\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m   File \"python/ray/_raylet.pyx\", line 824, in ray._raylet.store_task_errors\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m   File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 638, in temporary_actor_method\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m     raise RuntimeError(\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m RuntimeError: The actor with name ImplicitFunc failed to import on the worker. This may be because needed library dependencies are not installed in the worker environment:\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m \n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m Traceback (most recent call last):\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m   File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 677, in _load_actor_class_from_gcs\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m     actor_class = pickle.loads(pickled_class)\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m   File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/cloudpickle/cloudpickle.py\", line 665, in subimport\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m     __import__(name)\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m ModuleNotFoundError: No module named 'src'\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m \n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m \n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m During handling of the above exception, another exception occurred:\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m \n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m Traceback (most recent call last):\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m   File \"python/ray/_raylet.pyx\", line 1830, in ray._raylet.task_execution_handler\n",
      "\u001b[2m\u001b[36m(TemporaryActor pid=83511)\u001b[0m SystemExit\n",
      "2024-02-28 19:26:45,532\tWARNING worker.py:2006 -- Traceback (most recent call last):\n",
      "  File \"python/ray/_raylet.pyx\", line 1418, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1498, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1424, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1364, in ray._raylet.execute_task.function_executor\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 726, in actor_method_executor\n",
      "    return method(__ray_actor, *args, **kwargs)\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 638, in temporary_actor_method\n",
      "    raise RuntimeError(\n",
      "RuntimeError: The actor with name ImplicitFunc failed to import on the worker. This may be because needed library dependencies are not installed in the worker environment:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 677, in _load_actor_class_from_gcs\n",
      "    actor_class = pickle.loads(pickled_class)\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/cloudpickle/cloudpickle.py\", line 665, in subimport\n",
      "    __import__(name)\n",
      "ModuleNotFoundError: No module named 'src'\n",
      "\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"python/ray/_raylet.pyx\", line 1787, in ray._raylet.task_execution_handler\n",
      "  File \"python/ray/_raylet.pyx\", line 1684, in ray._raylet.execute_task_with_cancellation_handler\n",
      "  File \"python/ray/_raylet.pyx\", line 1366, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1367, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1583, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 824, in ray._raylet.store_task_errors\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 638, in temporary_actor_method\n",
      "    raise RuntimeError(\n",
      "RuntimeError: The actor with name ImplicitFunc failed to import on the worker. This may be because needed library dependencies are not installed in the worker environment:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 677, in _load_actor_class_from_gcs\n",
      "    actor_class = pickle.loads(pickled_class)\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/cloudpickle/cloudpickle.py\", line 665, in subimport\n",
      "    __import__(name)\n",
      "ModuleNotFoundError: No module named 'src'\n",
      "\n",
      "An unexpected internal error occurred while the worker was executing a task.\n",
      "2024-02-28 19:26:45,533\tWARNING worker.py:2006 -- A worker died or was killed while executing a task by an unexpected system error. To troubleshoot the problem, check the logs for the dead worker. RayTask ID: ffffffffffffffff2a0e3f4a8450157c0c9dcf6901000000 Worker ID: 9e3d4367cf03286c125e9c65871c5d4c4dc8f9bc0b003199daf4c827 Node ID: efc15fd61ccb1fcdfd79f1e80cc4e7e60403d12d4dea718b61bee81e Worker IP address: 127.0.0.1 Worker port: 60277 Worker PID: 83512 Worker exit type: SYSTEM_ERROR Worker exit detail: Worker exits unexpectedly. Worker exits with an exit code None.\n",
      " Traceback (most recent call last):\n",
      "  File \"python/ray/_raylet.pyx\", line 1418, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1498, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1424, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1364, in ray._raylet.execute_task.function_executor\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 726, in actor_method_executor\n",
      "    return method(__ray_actor, *args, **kwargs)\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 638, in temporary_actor_method\n",
      "    raise RuntimeError(\n",
      "RuntimeError: The actor with name ImplicitFunc failed to import on the worker. This may be because needed library dependencies are not installed in the worker environment:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 677, in _load_actor_class_from_gcs\n",
      "    actor_class = pickle.loads(pickled_class)\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/cloudpickle/cloudpickle.py\", line 665, in subimport\n",
      "    __import__(name)\n",
      "ModuleNotFoundError: No module named 'src'\n",
      "\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"python/ray/_raylet.pyx\", line 1787, in ray._raylet.task_execution_handler\n",
      "  File \"python/ray/_raylet.pyx\", line 1684, in ray._raylet.execute_task_with_cancellation_handler\n",
      "  File \"python/ray/_raylet.pyx\", line 1366, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1367, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1583, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 824, in ray._raylet.store_task_errors\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 638, in temporary_actor_method\n",
      "    raise RuntimeError(\n",
      "RuntimeError: The actor with name ImplicitFunc failed to import on the worker. This may be because needed library dependencies are not installed in the worker environment:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 677, in _load_actor_class_from_gcs\n",
      "    actor_class = pickle.loads(pickled_class)\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/cloudpickle/cloudpickle.py\", line 665, in subimport\n",
      "    __import__(name)\n",
      "ModuleNotFoundError: No module named 'src'\n",
      "\n",
      "An unexpected internal error occurred while the worker was executing a task.\n",
      "2024-02-28 19:26:45,534\tERROR tune_controller.py:911 -- Trial task failed for trial train_tune_2b9e5c13\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/air/execution/_internal/event_manager.py\", line 110, in resolve_future\n",
      "    result = ray.get(future)\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/auto_init_hook.py\", line 24, in auto_init_wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/client_mode_hook.py\", line 103, in wrapper\n",
      "    return func(*args, **kwargs)\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/worker.py\", line 2495, in get\n",
      "    raise value\n",
      "ray.exceptions.RayActorError: The actor died unexpectedly before finishing this task.\n",
      "\tclass_name: ImplicitFunc\n",
      "\tactor_id: 2a0e3f4a8450157c0c9dcf6901000000\n",
      "\tpid: 83512\n",
      "\tnamespace: 8e0d37a8-9651-4e34-a440-6bd4d070a181\n",
      "\tip: 127.0.0.1\n",
      "The actor is dead because its worker process has died. Worker exit type: SYSTEM_ERROR Worker exit detail: Worker exits unexpectedly. Worker exits with an exit code None.\n",
      " Traceback (most recent call last):\n",
      "  File \"python/ray/_raylet.pyx\", line 1418, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1498, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1424, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1364, in ray._raylet.execute_task.function_executor\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 726, in actor_method_executor\n",
      "    return method(__ray_actor, *args, **kwargs)\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 638, in temporary_actor_method\n",
      "    raise RuntimeError(\n",
      "RuntimeError: The actor with name ImplicitFunc failed to import on the worker. This may be because needed library dependencies are not installed in the worker environment:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 677, in _load_actor_class_from_gcs\n",
      "    actor_class = pickle.loads(pickled_class)\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/cloudpickle/cloudpickle.py\", line 665, in subimport\n",
      "    __import__(name)\n",
      "ModuleNotFoundError: No module named 'src'\n",
      "\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"python/ray/_raylet.pyx\", line 1787, in ray._raylet.task_execution_handler\n",
      "  File \"python/ray/_raylet.pyx\", line 1684, in ray._raylet.execute_task_with_cancellation_handler\n",
      "  File \"python/ray/_raylet.pyx\", line 1366, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1367, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1583, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 824, in ray._raylet.store_task_errors\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 638, in temporary_actor_method\n",
      "    raise RuntimeError(\n",
      "RuntimeError: The actor with name ImplicitFunc failed to import on the worker. This may be because needed library dependencies are not installed in the worker environment:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 677, in _load_actor_class_from_gcs\n",
      "    actor_class = pickle.loads(pickled_class)\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/cloudpickle/cloudpickle.py\", line 665, in subimport\n",
      "    __import__(name)\n",
      "ModuleNotFoundError: No module named 'src'\n",
      "\n",
      "An unexpected internal error occurred while the worker was executing a task.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div class=\"trialProgress\">\n",
       "  <h3>Trial Progress</h3>\n",
       "  <table>\n",
       "<thead>\n",
       "<tr><th>Trial name         </th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr><td>train_tune_2b9e5c13</td></tr>\n",
       "</tbody>\n",
       "</table>\n",
       "</div>\n",
       "<style>\n",
       ".trialProgress {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "  color: var(--jp-ui-font-color1);\n",
       "}\n",
       ".trialProgress h3 {\n",
       "  font-weight: bold;\n",
       "}\n",
       ".trialProgress td {\n",
       "  white-space: nowrap;\n",
       "}\n",
       "</style>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-02-28 19:26:46,869\tWARNING worker.py:2006 -- Traceback (most recent call last):\n",
      "  File \"python/ray/_raylet.pyx\", line 1418, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1498, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1424, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1364, in ray._raylet.execute_task.function_executor\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 726, in actor_method_executor\n",
      "    return method(__ray_actor, *args, **kwargs)\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 638, in temporary_actor_method\n",
      "    raise RuntimeError(\n",
      "RuntimeError: The actor with name ImplicitFunc failed to import on the worker. This may be because needed library dependencies are not installed in the worker environment:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 677, in _load_actor_class_from_gcs\n",
      "    actor_class = pickle.loads(pickled_class)\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/cloudpickle/cloudpickle.py\", line 665, in subimport\n",
      "    __import__(name)\n",
      "ModuleNotFoundError: No module named 'src'\n",
      "\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"python/ray/_raylet.pyx\", line 1787, in ray._raylet.task_execution_handler\n",
      "  File \"python/ray/_raylet.pyx\", line 1684, in ray._raylet.execute_task_with_cancellation_handler\n",
      "  File \"python/ray/_raylet.pyx\", line 1366, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1367, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1583, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 824, in ray._raylet.store_task_errors\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 638, in temporary_actor_method\n",
      "    raise RuntimeError(\n",
      "RuntimeError: The actor with name ImplicitFunc failed to import on the worker. This may be because needed library dependencies are not installed in the worker environment:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 677, in _load_actor_class_from_gcs\n",
      "    actor_class = pickle.loads(pickled_class)\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/cloudpickle/cloudpickle.py\", line 665, in subimport\n",
      "    __import__(name)\n",
      "ModuleNotFoundError: No module named 'src'\n",
      "\n",
      "An unexpected internal error occurred while the worker was executing a task.\n",
      "2024-02-28 19:26:46,871\tWARNING worker.py:2006 -- A worker died or was killed while executing a task by an unexpected system error. To troubleshoot the problem, check the logs for the dead worker. RayTask ID: ffffffffffffffffde9630ad80f8f8a5f93a08d301000000 Worker ID: 6f7322e054611344dc82e416620ed8b700dcf648f3b4da39e72bee79 Node ID: efc15fd61ccb1fcdfd79f1e80cc4e7e60403d12d4dea718b61bee81e Worker IP address: 127.0.0.1 Worker port: 60287 Worker PID: 83515 Worker exit type: SYSTEM_ERROR Worker exit detail: Worker exits unexpectedly. Worker exits with an exit code None.\n",
      " Traceback (most recent call last):\n",
      "  File \"python/ray/_raylet.pyx\", line 1418, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1498, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1424, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1364, in ray._raylet.execute_task.function_executor\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 726, in actor_method_executor\n",
      "    return method(__ray_actor, *args, **kwargs)\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 638, in temporary_actor_method\n",
      "    raise RuntimeError(\n",
      "RuntimeError: The actor with name ImplicitFunc failed to import on the worker. This may be because needed library dependencies are not installed in the worker environment:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 677, in _load_actor_class_from_gcs\n",
      "    actor_class = pickle.loads(pickled_class)\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/cloudpickle/cloudpickle.py\", line 665, in subimport\n",
      "    __import__(name)\n",
      "ModuleNotFoundError: No module named 'src'\n",
      "\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"python/ray/_raylet.pyx\", line 1787, in ray._raylet.task_execution_handler\n",
      "  File \"python/ray/_raylet.pyx\", line 1684, in ray._raylet.execute_task_with_cancellation_handler\n",
      "  File \"python/ray/_raylet.pyx\", line 1366, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1367, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1583, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 824, in ray._raylet.store_task_errors\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 638, in temporary_actor_method\n",
      "    raise RuntimeError(\n",
      "RuntimeError: The actor with name ImplicitFunc failed to import on the worker. This may be because needed library dependencies are not installed in the worker environment:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 677, in _load_actor_class_from_gcs\n",
      "    actor_class = pickle.loads(pickled_class)\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/cloudpickle/cloudpickle.py\", line 665, in subimport\n",
      "    __import__(name)\n",
      "ModuleNotFoundError: No module named 'src'\n",
      "\n",
      "An unexpected internal error occurred while the worker was executing a task.\n",
      "2024-02-28 19:26:46,872\tERROR tune_controller.py:911 -- Trial task failed for trial train_tune_6b12b628\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/air/execution/_internal/event_manager.py\", line 110, in resolve_future\n",
      "    result = ray.get(future)\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/auto_init_hook.py\", line 24, in auto_init_wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/client_mode_hook.py\", line 103, in wrapper\n",
      "    return func(*args, **kwargs)\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/worker.py\", line 2495, in get\n",
      "    raise value\n",
      "ray.exceptions.RayActorError: The actor died unexpectedly before finishing this task.\n",
      "\tclass_name: ImplicitFunc\n",
      "\tactor_id: de9630ad80f8f8a5f93a08d301000000\n",
      "\tpid: 83515\n",
      "\tnamespace: 8e0d37a8-9651-4e34-a440-6bd4d070a181\n",
      "\tip: 127.0.0.1\n",
      "The actor is dead because its worker process has died. Worker exit type: SYSTEM_ERROR Worker exit detail: Worker exits unexpectedly. Worker exits with an exit code None.\n",
      " Traceback (most recent call last):\n",
      "  File \"python/ray/_raylet.pyx\", line 1418, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1498, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1424, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1364, in ray._raylet.execute_task.function_executor\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 726, in actor_method_executor\n",
      "    return method(__ray_actor, *args, **kwargs)\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 638, in temporary_actor_method\n",
      "    raise RuntimeError(\n",
      "RuntimeError: The actor with name ImplicitFunc failed to import on the worker. This may be because needed library dependencies are not installed in the worker environment:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 677, in _load_actor_class_from_gcs\n",
      "    actor_class = pickle.loads(pickled_class)\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/cloudpickle/cloudpickle.py\", line 665, in subimport\n",
      "    __import__(name)\n",
      "ModuleNotFoundError: No module named 'src'\n",
      "\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"python/ray/_raylet.pyx\", line 1787, in ray._raylet.task_execution_handler\n",
      "  File \"python/ray/_raylet.pyx\", line 1684, in ray._raylet.execute_task_with_cancellation_handler\n",
      "  File \"python/ray/_raylet.pyx\", line 1366, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1367, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 1583, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 824, in ray._raylet.store_task_errors\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 638, in temporary_actor_method\n",
      "    raise RuntimeError(\n",
      "RuntimeError: The actor with name ImplicitFunc failed to import on the worker. This may be because needed library dependencies are not installed in the worker environment:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/_private/function_manager.py\", line 677, in _load_actor_class_from_gcs\n",
      "    actor_class = pickle.loads(pickled_class)\n",
      "  File \"/Users/samirgupta/Desktop/WORK/--_work/Research/After_12th_December_2023/05_GNN_occu_prediction/WeGo_Occu_pred/NEW_FOLDER/CODE/venv/lib/python3.10/site-packages/ray/cloudpickle/cloudpickle.py\", line 665, in subimport\n",
      "    __import__(name)\n",
      "ModuleNotFoundError: No module named 'src'\n",
      "\n",
      "An unexpected internal error occurred while the worker was executing a task.\n"
     ]
    }
   ],
   "source": [
    "route_info = {} \n",
    "\n",
    "route_info['route_id'] = 55\n",
    "route_info['num_classes'] = 4\n",
    "\n",
    "\n",
    "ray_hyper_parameter_tune(route_info)\n",
    "\n",
    "ray.shutdown()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
